An alternative algo for generating initial meshes

vovannikov opened the pull request:
This should render more consistent number of DOFs. I will post here a few tests.

vovannikov commented:
I modified the previous algorithm for decomposing the divisions number into prime numbers and it is now inside function `adjust_divisions_to_primes()`. The new version generates the following numbers of DoFs:

|  **# particles**  |   **51**  |  **102**  |  **212** |  **316**  |  **603**  |  **1037**  |  **3076**  |
|:-----------------:|:---------:|:---------:|:--------:|:---------:|:---------:|:---------:|:---------:|
| **# scalar dofs** | 3,354,187 | 6,362,009 | 11,704,953 | 20,931,860 | 37,864,506 | 59,510,435 | 170,211,814 |
| **# dofs / particle** | 65,768 | 62,373 | 55,212 | 66,240 | 62,793 | 57,387 | 55,335 |

![image](https://github.com/hpsint/hpsint/assets/8836201/b214a5b2-93fc-4a8f-ab91-80c8498b7c2c)

I also added an alternative algorithm for the initial meshes but it generates slightly less consistent dofs numbers. I decided to keep for future tests if we encounter some further problems.

I also added some test for keeping future track in the meshing algos.

In the context of the paper, we would need to pick just a couple of commits from this PR, namely:
ede8b9a1e91fd815aa2ff547fe401a660c639179 fc8276d28840aa5d1bd42b289cceb907e2e72992 bd417c215a27a463ce5896af2992c5b77329134e d5182226e5f6ea91536564d4a706dc5d36c1cabc (not mandatory) 7df4ddf2c9b57792e640789b39c260b3ebf1db9d (not mandatory)

peterrum commented:
> I modified the previous algorithm for decomposing the divisions number into prime numbers and it is now inside function adjust_divisions_to_primes(). The new version generates the following numbers of DoFs:

Have you also checked if the number of coarse cells is reasonable?

peterrum commented:
> It seems that tests get dependent on the deal.II version or on the platform. @peterrum, is there something I am missing that can result in such behavior? See https://github.com/hpsint/hpsint/actions/runs/6026560741/job/16349792783?pr=573, slightly different numbers of dofs are generated for a single test:
> 
> ```
> tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Source: /__w/hpsint/hpsint/tests/initial_mesh_102p_from_interface.output
> tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Result: /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/output
> Check /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/output /__w/hpsint/hpsint/tests/initial_mesh_102p_from_interface.output
> tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Diff:   /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/failing_diff
> tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ First 20 lines of numdiff output:
> ----------------
> ##8       #:3   <== 3[61](https://github.com/hpsint/hpsint/actions/runs/6026560741/job/16349792783?pr=573#step:5:62)71
> ##8       #:3   ==> 36178
> @ Absolute error = 7.0000000000e+0, Relative error = 1.9352519975e-4
> ----------------
> ##9       #:3   <== 224275
> ##9       #:3   ==> 224279
> @ Absolute error = 4.0000000000e+0, Relative error = 1.7835246907e-5
> ```

Maybe you are using different boost versions?

peterrum commented on applications/sintering/include/pf-applications/sintering/tools.h:
```diff
@@ -144,19 +151,102 @@ namespace Sintering
-              const double        max_level0_divisions_per_interface = 1.0,
-              const unsigned int  divisions_per_element              = 1,
-              const bool          print_stats                        = true)
+  create_mesh_from_interface(
```
Could you add a comment describing what each function does?

vovannikov replied:
Yep, done!

vovannikov commented:
> > I modified the previous algorithm for decomposing the divisions number into prime numbers and it is now inside function adjust_divisions_to_primes(). The new version generates the following numbers of DoFs:
> 
> Have you also checked if the number of coarse cells is reasonable?

I can probably add this checks also to the tests. Besides that, I also added a test that checks the decomposition of the cells with prime numbers, it covers some aspects of this check (https://github.com/hpsint/hpsint/blob/e48d34636e89cde4088bd979d6aacd10fa61c744/tests/adjust_divisions.cc).

vovannikov commented:
> > It seems that tests get dependent on the deal.II version or on the platform. @peterrum, is there something I am missing that can result in such behavior? See https://github.com/hpsint/hpsint/actions/runs/6026560741/job/16349792783?pr=573, slightly different numbers of dofs are generated for a single test:
> > ```
> > tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Source: /__w/hpsint/hpsint/tests/initial_mesh_102p_from_interface.output
> > tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Result: /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/output
> > Check /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/output /__w/hpsint/hpsint/tests/initial_mesh_102p_from_interface.output
> > tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ Diff:   /__w/hpsint/hpsint/build/tests/initial_mesh_102p_from_interface.release/failing_diff
> > tests/initial_mesh_102p_from_interface.release: DIFF failed. ------ First 20 lines of numdiff output:
> > ----------------
> > ##8       #:3   <== 3[61](https://github.com/hpsint/hpsint/actions/runs/6026560741/job/16349792783?pr=573#step:5:62)71
> > ##8       #:3   ==> 36178
> > @ Absolute error = 7.0000000000e+0, Relative error = 1.9352519975e-4
> > ----------------
> > ##9       #:3   <== 224275
> > ##9       #:3   ==> 224279
> > @ Absolute error = 4.0000000000e+0, Relative error = 1.7835246907e-5
> > ```
> 
> Maybe you are using different boost versions?

This could be the case since I used my system's boost when compiling deal.II lately. I will force to use the bundled and then report.

vovannikov commented:
> > Maybe you are using different boost versions?
> 
> This could be the case since I used my system's boost when compiling deal.II lately. I will force to use the bundled and then report.

Unfortunately, that did not help. My numbers have not changed.
